src/shell/interpreter.zig

```diff
@@ -3054,6 +3139,591 @@ pub const Interpreter = struct {
+                switch (this.state) {
+                    .idle => {
+                        this.state = .{ .expanding_args = .{ .expansion = undefined } };
+                        continue;
+                    },
+                    .expanding_args => {
+                        if (this.state.expanding_args.idx >= this.node.args.len()) {
+                            this.commandImplStart();
+                            return;
+                        }
+
+                        this.args.ensureUnusedCapacity(1) catch bun.outOfMemory();
+                        Expansion.init(
+                            this.base.interpreter,
+                            this.base.shell,
+                            &this.state.expanding_args.expansion,
+                            this.node.args.getConst(this.state.expanding_args.idx),
+                            Expansion.ParentPtr.init(this),
+                            .{
+                                .array_of_slice = &this.args,
+                            },
+                            this.io.copy(),
+                        );
+                        this.state.expanding_args.idx += 1;
+                        this.state.expanding_args.expansion.start();
+                        return;
+                    },
+                    .waiting_stat => return,
+                    .stat_complete => {
+                        switch (this.node.op) {
+                            .@"-f" => {
+                                this.parent.childDone(this, if (this.state.stat_complete.stat == .result) 0 else 1);
+                                return;
+                            },
+                            .@"-d" => {
+                                const st: bun.Stat = switch (this.state.stat_complete.stat) {
+                                    .result => |st| st,
+                                    .err => {
+                                        // It seems that bash always gives exit code 1
+                                        this.parent.childDone(this, 1);
+                                        return;
+                                    },
+                                };
+                                this.parent.childDone(this, if (bun.S.ISDIR(@intCast(st.mode))) 0 else 1);
+                                return;
+                            },
+                            .@"-c" => {
+                                const st: bun.Stat = switch (this.state.stat_complete.stat) {
+                                    .result => |st| st,
+                                    .err => {
+                                        // It seems that bash always gives exit code 1
+                                        this.parent.childDone(this, 1);
+                                        return;
+                                    },
+                                };
+                                this.parent.childDone(this, if (bun.S.ISCHR(@intCast(st.mode))) 0 else 1);
```

Comment:
Why are we doing special checks for ISCHR? A file, pipe, a socket can all be a TTY and not a character device. You most likely want isatty instead.

Reply:
This is for -c conditional expression which checks that the file is a character device